### PR TITLE
Automatically Update Data Options

### PR DESCRIPTION
**Pull request by Rockyd3**
Updated the plottr app to allow for continuous checking of everything within the directory the program is run from. If a new data directory is added, this automatically refreshes the options to show the new options.

Changed multi select widget to check box widget to better allow this. Made data options include dates to avoid duplicate file names getting lost.

to test: run 
panel serve [path to labcore\labcore\scripts\monitr-server.py]
in a directory that has data.

### CODE REVIEW

File: labcore/analysis/hvapps.py [hunk 1]
```diff
@@ -242,53 +350,64 @@ def __init__(self, *args: Any, **kwargs: Any):
...
     @pn.depends("refresh.value", watch=True)
     def on_refresh_changed(self):
+        # try:
+        #     # Remove the existing callback if it exists
+        #     pn.state.remove_periodic_callback(self.load_and_preprocess)
+        # except:
+        #     pass
+        # if self.refresh.value is not None:
+        #     # Add a new callback with different frequency
+        #     pn.state.add_periodic_callback(self.load_and_preprocess, self.refresh.value * 1000)
```
Comment (marcosfrenkel) [inline labcore/analysis/hvapps.py:410]: Remove commented code

File: labcore/analysis/hvplotting.py [hunk 2]
```diff
@@ -643,7 +643,11 @@ def plot_xr_as_2d(ds, x, y, dim_labels={}):
                     ylabel=dim_labels.get(y, y),
                     clabel=f"Mean {dim_labels.get(d, d)}",
                 )
-        return plot.cols(1)
+        # FIXME: QuadMesh object has no attribute 'cols' error for longsweep
+        try:
+            return plot.cols(1)
+        except:
```
Comment (marcosfrenkel) [inline labcore/analysis/hvplotting.py:649]: Add exception that you are trying to catch

**Review by marcosfrenkel — CHANGES_REQUESTED** [review]
Basic styling comments

File: labcore/analysis/hvapps.py [hunk 3]
```diff
@@ -2,14 +2,18 @@
...
 
 import asyncio
 import nest_asyncio
 nest_asyncio.apply()
 
+import pandas
 import param
 import panel as pn
 from panel.widgets import RadioButtonGroup as RBG, MultiSelect, Select
+from watchdog.observers import Observer
```
Comment (marcosfrenkel) [inline labcore/analysis/hvapps.py:15]: Add watchdog as a requirement to the package

File: labcore/analysis/hvapps.py [hunk 4]
```diff
@@ -22,6 +26,13 @@
 )
 from .hvplotting import Node, labeled_widget
 
+class Handler(FileSystemEventHandler):
```
Comment (marcosfrenkel) [inline labcore/analysis/hvapps.py:29]: 2 empty spaces between members (imports, functions, classes, etc)

File: labcore/analysis/hvapps.py [hunk 5]
```diff
@@ -22,6 +26,13 @@
...
+    def __init__(self, update_callback):
+        self.update_callback = update_callback
+
+    def on_created(self, event):
+        if event.is_directory:
+            self.update_callback(event)
```
Comment (marcosfrenkel) [inline labcore/analysis/hvapps.py:35]: Same, 2 spaces between this line and line underneath

File: labcore/analysis/hvapps.py [hunk 6]
```diff
@@ -84,48 +100,95 @@ def __init__(self, data_root, size=15):
             css_classes=['ttlabel'],
         )
         self.layout.append(self.text_input_repeater)
+        
+        
+        self.image_feed_width = 400  # The width of images in the feed
```
Comment (marcosfrenkel) [inline labcore/analysis/hvapps.py:105]: Single blank line between code inside a function

File: labcore/analysis/hvapps.py [hunk 7]
```diff
@@ -84,48 +100,95 @@ def __init__(self, data_root, size=15):
...
+        self.feed_scroll_width = 40  # Extra width of the feed itself for the scroll bar
 
         # two selectors for data selection
-        self._group_select_widget = MultiSelect(
+        self._group_select_widget = pn.widgets.CheckBoxGroup(
             name='Date', 
-            size=self.size,
-            width=200,
+            # height=self.size,
```
Comment (marcosfrenkel) [inline labcore/analysis/hvapps.py:111]: Remove commented code

File: labcore/analysis/hvapps.py [hunk 8]
```diff
@@ -84,48 +100,95 @@ def __init__(self, data_root, size=15):
...
+            width=200-self.feed_scroll_width,
             stylesheets = [selector_stylesheet]
```
Comment (marcosfrenkel) [inline labcore/analysis/hvapps.py:113]: This is not on you, but can you remove the spaces between the `=`?

File: labcore/analysis/hvapps.py [hunk 9]
```diff
@@ -84,48 +100,95 @@ def __init__(self, data_root, size=15):
...
+        self.DIRECTORY_TO_WATCH = r"."
+        self.observer = Observer()
+        self.handler = Handler(self.update_group_options)
+        self.start()
+
+    def start(self):
+        self.observer.schedule(self.handler, self.DIRECTORY_TO_WATCH, recursive=True)
+        self.observer.start()
+
+
```
Comment (marcosfrenkel) [inline labcore/analysis/hvapps.py:171]: Too many empty lines between members of a class

**Review by marcosfrenkel — CHANGES_REQUESTED** [review]

**Comment by marcosfrenkel** [comment]
So the way the this is working at the moment is that at any change you reload the entire data? You are not filtering what the event is and modifying only the necessary thing right?

**Comment by Rockyd3** [comment]
> So the way the this is working at the moment is that at any change you reload the entire data? You are not filtering what the event is and modifying only the necessary thing right?

That is correct. There is probably a better way to do this - I'm not sure what parts of the way data is stored are consistent/need to be consistent. I will look into making this occur with the current assumed data structure/labelling

**Comment by marcosfrenkel** [comment]
We should have a little bit of a discussion in person about how to handle this. It might not be worth doing this now though, but focusing on other features before. Let's talk in person about this

**Review by marcosfrenkel — APPROVED** [review]
Looks good

**Pull request merged**